File: code/template/python3/Dinic.py
from collections import deque

class Dinic:
    def __init__(self, size):
        self.size = size
        self.edges = [[] for _ in range(size)]
        self.depth = None
        self.progress = None

    def addEdge(self, fr, to, cap):
        self.edges[fr].append([cap, to, len(self.edges[to])])
        self.edges[to].append([0, fr, len(self.edges[fr]) - 1])

    def bfs(self, s):
        depth = [-1] * self.size
        depth[s] = 0
        que = deque([s])

        while que:
            now = que.popleft()
            for cap, to, _ in self.edges[now]:
                if cap > 0 and depth[to] < 0:
                    depth[to] = depth[now] + 1
                    que.append(to)
        self.depth = depth

    def dfs(self, s, t, flow):
        if s == t:
            return flow

        for i, (cap, to, rev) in enumerate(self.edges[s][self.progress[s]:], start=self.progress[s]):
            self.progress[s] = i
            if cap == 0 or self.depth[s] >= self.depth[to]:
                continue
            d = self.dfs(to, t, min(flow, cap))
            if d == 0:
                continue
            self.edges[s][i][0] -= d
            self.edges[to][rev][0] += d
            return d
        return 0

    def maxFlow(self, s, t):
        flow = 0
        while True:
            self.bfs(s)
            if self.depth[t] < 0:
                return flow
            self.progress = [0] * self.size
            currentFlow = self.dfs(s, t, float('inf'))
            while currentFlow > 0:
                flow += currentFlow
                currentFlow = self.dfs(s, t, float('inf'))

File: code/template/c++/Dinic.cpp
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;
typedef long long llong;
typedef vector<llong> vecllong;
typedef vector<vector<llong>> vvecllong;

const llong MOD = 1e9 + 7;
const llong INF = 1e17;

#define FOR(i, n) for (llong i = 0; i < n; i++)
#define FORS(i, a, b) for (llong i = a; i < b; i++)
#define FORR(i, n) for (llong i = n; i > 0; i++)

struct edge {
    llong capacity;
    llong to;
    llong reverseEdge;
};

class Dinic {
    private:
        llong size;
        vector<vector<edge>> edges;
        vecllong depth;
        vecllong progress;

        void bfs(llong s) {
            vecllong depth(this->size, -1);
            depth[s] = 0;
            deque<llong> que(1, s);

            while (!que.empty()) {
                llong now = que.front();
                que.pop_front();
                for_each(this->edges[now].begin(), this->edges[now].end(), [&](edge e) {
                    if (e.capacity > 0 && depth[e.to] < 0) {
                        depth[e.to] = depth[now] + 1;
                        que.push_back(e.to);
                    }
                });
            }
            this->depth = depth;
        }

        llong dfs(llong s, llong t, llong flow) {
            if (s == t) return flow;
            vector<edge> es = this->edges[s];
            FORS(i, this->progress[s], es.size()) {
                this->progress[s] = i;
                edge e = es[i];
                if (e.capacity == 0 || this->depth[s] >= this->depth[e.to]) continue;

                llong d = this->dfs(e.to, t, min(flow, e.capacity));

                if (d == 0) continue;

                this->edges[s][i].capacity -= d;
                this->edges[e.to][e.reverseEdge].capacity += d;

                return d;
            }
            return 0;
        };

    public:
        Dinic (llong size) {
            this->init(size);
        };

        void init(llong size) {
            this->size = size;
            this->edges = vector<vector<edge>>(size, vector<edge>(0));
        };

        void addEdge(llong fr, llong to, llong cap) {
            this->edges[fr].push_back({cap, to, this->edges[to].size()});
            this->edges[to].push_back({0, fr, this->edges[fr].size() - 1});
        };

        llong maxFlow(llong s, llong t) {
            llong flow = 0;
            while (true) {
                this->bfs(s);
                if (this->depth[t] < 0) return flow;
                this->progress = vecllong(this->size, 0);
                llong currentFlow = this->dfs(s, t, INF);
                while (currentFlow > 0) {
                    flow += currentFlow;
                    currentFlow = this->dfs(s, t, INF);
                }
            }
        }
};
